src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -518,28 +518,23 @@ static int uct_tcp_sockcm_ep_send_skip_event(uct_tcp_sockcm_ep_t *cep)
         return cep->state & UCT_TCP_SOCKCM_EP_DATA_SENT;
     } else {
         ucs_assert(cep->state & UCT_TCP_SOCKCM_EP_ON_CLIENT);
-        return (cep->state & UCT_TCP_SOCKCM_EP_CLIENT_NOTIFY_SENT) ||
-               ((cep->state & UCT_TCP_SOCKCM_EP_DATA_SENT) &&
-                !(cep->state & UCT_TCP_SOCKCM_EP_CLIENT_NOTIFY_CALLED));
+        return (cep->state & (UCT_TCP_SOCKCM_EP_CLIENT_NOTIFY_SENT |
+                              UCT_TCP_SOCKCM_EP_DATA_SENT)) ||
+               !(cep->state & UCT_TCP_SOCKCM_EP_PRIV_DATA_PACKED);
     }
 }
 
 ucs_status_t uct_tcp_sockcm_ep_send(uct_tcp_sockcm_ep_t *cep)
 {
-    ucs_status_t status;
-
-    if (uct_tcp_sockcm_ep_send_skip_event(cep)) {
-        return UCS_OK;
-    }
-
     if (!(cep->state & (UCT_TCP_SOCKCM_EP_RESOLVE_CB_INVOKED |
                         UCT_TCP_SOCKCM_EP_PRIV_DATA_PACKED   |
                         UCT_TCP_SOCKCM_EP_ON_SERVER))) {
         ucs_assert(cep->state & UCT_TCP_SOCKCM_EP_ON_CLIENT);
-        status = uct_tcp_sockcm_ep_resolve(cep);
-        if (status != UCS_OK) {
-            return status;
-        }
+        return uct_tcp_sockcm_ep_resolve(cep);
+    }
+
+    if (uct_tcp_sockcm_ep_send_skip_event(cep)) {
```

Comment:
why is this condition moved to here? what if UCT_TCP_SOCKCM_EP_DISCONNECTING is set?

Reply:
do you mean `UCT_TCP_SOCKCM_EP_DISCONNECTING` is set? it's impossible since `uct_ep_disconnect` should return `BUSY` if the ep is not connected yet, i.e. resolve CB must be called earlier

Reply:
@evgeny-leksikov what if the ep was connected, then `uct_ep_disconnect` was called, this flag was set and then 'an earlier queued' write event resulted in calling this function?
this check was added there since, AFAIR, there was some race where i saw this. it was a long time ago so i dont have the details of the case at the moment...

Reply:
IMO, if we have 'an earlier queued' write event this means we should have UCT_TCP_SOCKCM_EP_PRIV_DATA_PACKED flag (since could not send private data w/o packing it), so it would not go inside this if
we can add assert inside that if, that DISCONNECTING flag is not set

Reply:
>we can add assert inside that if, that DISCONNECTING flag is not set

👍